art/attacks/evasion/steal_now_attack_latter.py

```diff
@@ -0,0 +1,747 @@
+            for idx in range(n_samples):
+                cur_patch = adv_patch[idx]
+                cur_position = adv_position[idx]
+
+                for key, obj in tile_mat.items():
+
+                    ii, jj = key
+                    b1 = obj.xyxy
+                    obj_threshold = obj.threshold
+                    [x1, y1, x2, y2] = b1.type(torch.IntTensor)
+                    overlay = _bbox_ioa(b1.type(torch.FloatTensor), cur_position.type(torch.FloatTensor))
+                    bcount = torch.sum(overlay > 0.0).item()
+
+                    x_ref = x_org[:, :, y1:y2, x1:x2]
+                    x_cur = x_cand[idx, :, y1:y2, x1:x2].clone()
+
+                    pert = x_cur - x_ref
+                    loss = self._get_loss(pert, self.eps)
+                    eligible = torch.max(torch.abs(pert)) < self.eps and bcount >= obj_threshold
+                    TPatch_cur = TileObj(tile_size=tile_size, device=self.estimator.model.device)
+                    TPatch_cur.update(eligible, bcount, torch.sum(loss), x_cur)
+                    # insert op
+                    prev = tile_mat[(ii, jj)]
+                    prev.insert(TPatch_cur)
+                    tile_mat[(ii, jj)] = prev
+
+        # clean non-active regions
+        x_out = x_init.clone()
+        x_eval = self._assemble(tile_mat, x_org)
+        adv_patch, adv_position = collect_patches_from_images(self.estimator, x_eval)
+        cur_patch = adv_patch[0]
+        cur_position = adv_position[0]
+        for key, obj in tile_mat.items():
+            ii, jj = key
+            b1 = obj.xyxy
+            [x1, y1, x2, y2] = b1.type(torch.IntTensor)
+            overlay = _bbox_ioa(b1.type(torch.FloatTensor), cur_position.type(torch.FloatTensor))
+            bcount = torch.sum(overlay > 0.0).item()
+
+            x_ref = x_init[:, :, y1:y2, x1:x2]
+            x_tag = x_eval[:, :, y1:y2, x1:x2]
+            cur_mask = torch.zeros_like(x_ref)
+            if bcount > 1:
+                bbox = cur_position[overlay > 0.0]
+                for b in bbox:
+                    bx1 = torch.clamp_min(b[0] - x1, 0)
+                    by1 = torch.clamp_min(b[1] - y1, 0)
+                    bx2 = torch.clamp_max(b[2] - x1, (x2 - x1 - 1).to(self.estimator.model.device))
+                    by2 = torch.clamp_max(b[3] - y1, (y2 - y1 - 1).to(self.estimator.model.device))
+                    cur_mask[:, :, by1:by2, bx1:bx2] = 1.0
+            else:
+                prev = tile_mat[(ii, jj)]
+                prev.pop()
+                tile_mat[(ii, jj)] = prev
+
+            a_mask = _drop_block2d(x_ref, 0.05, 1)
+            cur_mask = cur_mask * a_mask
+            updated = ((1.0 - cur_mask)  * x_ref) + cur_mask * ( 0.0 * x_ref + 1.0 * x_tag)
+            updated = ((1.0 - cur_mask)  * x_ref) + cur_mask * ( 0.0 * x_ref + 1.0 * updated)
+
+            x_out[:, :, y1:y2, x1:x2] = updated
+
+        return x_out, tile_mat
```

Comment:
Add empty line at end of file
```suggestion
        return x_out, tile_mat

```